### 1

a)

In [17]:
# tabela
x = [0.8,   1,   2,   3,   4]
y = [2.4, 2.5, 2.8, 3.1, 3.4]

Linearizando

$$
y = A \frac{1}{x^m} \\
\frac{1}{y} = \frac{1}{A} x^m \\
\ln \left( \frac{1}{y} \right) = \ln \left(  \frac{1}{A} x^m \right) \\
\ln \left( \frac{1}{y} \right) = \ln \left(  \frac{1}{A} \right) + m\ln(x) \\
$$

É preciso resolver
$$
\phi^T \phi a = \phi^T y
$$

$$
\phi = \left[\begin{matrix}1.0 & -0.22314355131421\\1.0 & 0.0\\1.0 & 0.693147180559945\\1.0 & 1.09861228866811\\1.0 & 1.38629436111989\end{matrix}\right] \quad
y = \left[\begin{matrix}-0.8754687373539\\-0.916290731874155\\-1.02961941718116\\-1.1314021114911\\-1.22377543162212\end{matrix}\right] \quad
\beta = \left[\begin{matrix}\beta_1\\\beta_2\end{matrix}\right]
$$

Onde $\phi$ é a matriz de Vandermonde

In [14]:
import sympy as sp
from math import log

def equacoes_normais(lista_x, lista_y, grau):
    vandermonde = [[x**p for p in range(grau+1)] for x in lista_x] # cria a matriz de Vandermonde
    
    phi = sp.Matrix(vandermonde)
    y = sp.Matrix(lista_y) # matriz coluna dos valores de f(x)
    
    A = phi.T * phi
    B = phi.T * y
    
    return A.gauss_jordan_solve(B)[0] # resolve Ax = B e retorna x

equacoes_normais(lista_x=[log(x) for x in x], lista_y=[log(1/y) for y in y], grau=1)

Matrix([
[-0.912166845959462],
[-0.208372553337376]])

$$
m = -0.208 \\
\ln \left(  \frac{1}{A} \right) = -0.912 \\
e^{\ln \left(  \frac{1}{A} \right)} = e^{-0.912} \\
\frac{1}{A} = e^{-0.912} \\
A = \frac{1}{e^{-0.912}} \\
A = 2.489
$$

Então
$$
y = 2.489 \frac{1}{x^{-0.208}}
$$

b) $a e^{b \sin(x) + cx^2}$

Linearizando

$$
y = a e^{b \sin(x) + cx^2} \\
\ln(y) = \ln(a e^{b \sin(x) + cx^2}) \\
\ln(y) = \ln(a) + \ln(e^{b \sin(x) + cx^2}) \\
\ln(y) = \ln(a) + b \sin(x) + cx^2 \\
$$

In [16]:
from math import sin
equacoes_normais(lista_x=[sin(x) for x in x], lista_y=[log(1/y) for y in y], grau=2)

Matrix([
[ -1.11064943916063],
[ 0.175391180412714],
[0.0196009239601559]])

$$
\ln(a) = -1.11 \quad a = e^{-1.11} \quad a = 0.33 \\
b = 0.175 \\
c = 0.020 
$$

$$
0.33 e^{0.175 \sin(x) + 0.02x^2}
$$

### 2)

In [4]:
f = lambda x,y: 3*y - x 

Usando o método de Euler

In [6]:
from numpy import arange

def euler(x0, x_f, h, f, y0):
    resultado = y0
    for x in arange(x0, x_f, h):
        resultado = resultado + h * f(x, resultado)
        
    return resultado

In [7]:
euler(0, 1, 0.2, f, 0.1)

0.32793600000000006

a) 0.3279

In [8]:
euler(0, 1, 0.5, f, 0.1)

0.375

b) 0.375

c)

In [11]:
abs((0.327936 - 0.375) / 0.327936)

0.14351580796252925

O erro relativo é de 0.14351

### 3

a)

Com 5 pontos, temos 4 intervalos. h = 0.25

$$
(f(0) + 4 f(0.25) + 2 f(0.5) + 4 f(0.75) + f(1)) \cdot \frac{0.25}{3} \\
(1.0 + 4 \cdot 1.06449 + 2 \cdot 1.28403 + 4 \cdot 1.75505 + 2.71828) \cdot \frac{0.25}{3} \\
(1.0 + 4.25798 + 2.56805 + 7.02022 + 2.71828) \cdot 0.08333 = 1.46371
$$

b)

$$
R_T(n) \leq \frac{M (b - a)^3}{12n^2}
$$
Onde $M = \mathrm{max} |f''(x)|$ para $x \in [0, 1]$ e $n$ é o número de intervalos.

$$
f''(x) = 2 e^{(x^2)} (1 + 2 x^2)
$$
Então x = 1
$$
M = 6e
$$

$$
10^{-2} \geq \frac{6e (1 - 0)^3}{12n^2} \\
10^{-2} \geq \frac{e}{2n^2} \\
n^2 \geq \frac{e}{10^{-2}2} \\
n^2 \geq 136 \\
n \geq 11.66 \\
n = 12
$$

In [3]:
from numpy import linspace
def regra_trapezio(f, a, b, intervalos):
    soma = f(a) + f(b) # inicializa a soma com o valor de f(a) + f(b)
    
    for x in linspace(a, b, intervalos+1)[1:-1]: # divide [a, b] em intervalos
        soma += 2 * f(x)
    
    delta_x = (b-a)/intervalos
    
    return (delta_x/2) * soma

from math import exp

f = lambda x: exp(x**2)
regra_trapezio(f, 0, 1, 12)

1.4657942734011131

$$
I = 1.4657942734011131
$$

c)

$$
10^{-4} \geq \frac{6e (1 - 0)^3}{12n^2} \\
10^{-4} \geq \frac{e}{2n^2} \\
n^2 \geq \frac{e}{10^{-4}2} \\
n^2 \geq 13591.41 \\
n \geq 116.58 \\
$$

117 pontos, então.

### 4)

a)

In [1]:
# Tabela
x = [0.5, 1.2, 2.3, 3.4, 4.5]
y = [3.4, 4.5, 2.8, 1.1, 6.4]

$$
\frac{x-2.3}{1.2-2.3} \cdot \frac{x-3.4}{1.2-3.4} \cdot 4.5 + \frac{x-1.2}{2.3-1.2} \cdot \frac{x-3.4}{2.3-3.4} \cdot 2.8 + \frac{x-1.2}{3.4-1.2} \cdot \frac{x-2.3}{3.4-2.3} \cdot 1.1 \\
1.86x^2 - 10.599x + 14.541 + -2.314x^2 + 10.645x - 9.441 + 0.455x^2 - 1.591x + 1.255 \\
6.355 - 1.545x
$$

In [2]:
(lambda x: 6.355 - 1.545*x)(2.5)

2.4925000000000006

$$
p(2.5) = 2.492
$$

b)

In [29]:
def diferencas_divididas(points):
    if len(points) == 1:
        return points[0][1]
    
    return (diferencas_divididas(points[1:]) - diferencas_divididas(points[:-1]))/(points[-1][0] - points[0][0])

In [36]:
x = [0.5, 1.2, 2.3, 3.4, 4.5]
y = [3.4, 4.5, 2.8, 1.1, 6.4]

pontos = [(x,y) for x,y in zip(x, y)]
print(pontos)
diferencas_divididas(pontos[1:])

[(0.5, 3.4), (1.2, 4.5), (2.3, 2.8), (3.4, 1.1), (4.5, 6.4)]


0.8765339343851739

$$
erro(2.5) \leq |(2.5-1.2)*(2.5-2.3)*(2.5-3.4)| * 0.8653 \\
erro(2.5) \leq 0.234 * 0.8653 \\
erro(2.5) \leq 0.20510
$$